### DISCUSSION TITLE
How do I support older browsers?

### DISCUSSION THREAD

**Discussion by qianjiachun** [remix-run/remix#3240] [Q&A] [answered]
The compiled code contains `??`, It cannot run in some older browsers.  
I try to modify the `target` in `tsconfig` and `package.json` add `browserslist`, but it seems to fail. After build `??` still exist.  
  
So is there any way to make the compiled code support the old version of the browser?

**Comment by kiliman** [comment, marked as answer]
Remix requires ESM module support on the browsers. So if you need to support older browsers, then you'll need to remove the `<Script/>` tag from *root*. Then treat your site as non-JS only. You can do that conditionally based on the user-agent.

https://caniuse.com/es6-module

**Reply by qianjiachun** [reply]
thanks

**Reply by sergiodxa** [reply]
You don't even need to remove the <Scripts /> component, because ESM scripts use `<script type="module">` any browser which doesn't support ESM will ignore that script tag behaving as if you removed the <Scripts /> component.

The only scenario where this could break is if there's some JS feature being used on the browser code but not supported on browsers which also support ESM.

**Reply by shuaixr** [reply]
In the official examples of material ui and remix : https://github.com/mui/material-ui/blob/b3548fabeab1980d3c5e5b85945550df95ee6b82/examples/remix-with-typescript/app/root.tsx#L18

If don't convert the ESM to CommonJS, the MUI will not work in any not support ESM browser

As long as the UI library uses emotion, similar code is required, and they cannot be used in not support ESM browsers

**Reply by qianjiachun** [reply]
> Remix requires ESM module support on the browsers. So if you need to support older browsers, then you'll need to remove the `<Script/>` tag from _root_. Then treat your site as non-JS only. You can do that conditionally based on the user-agent.
> 
> https://caniuse.com/es6-module

I'm sorry to ask you some questions again @kiliman 

https://caniuse.com/es6-module shows `chrome >= 61`
But the build file contains `??`. This operator support chrome >= 85.

https://caniuse.com/?search=%3F%3F

So...actually, the chrome supported by remix is 85 ?

**Reply by sergiodxa** [reply]
Is ?? used by Remix? Or maybe by React or another lib?

**Reply by machour** [reply]
@sergiodxa Seems like we have a few of those. Might be the one in `packages/remix-react/components.tsx` that's causing troubles.
It got introduced in #3697 

```
./integration/helpers/create-fixture.ts:139:  let template = init.template ?? "node-template";
./integration/helpers/create-fixture.ts:205:    Object.keys(init.files ?? {}).map(async (filename) => {
./docs/tutorials/jokes.md:2461:              searchParams.get("redirectTo") ?? undefined
./docs/tutorials/jokes.md:2671:              searchParams.get("redirectTo") ?? undefined
./docs/tutorials/jokes.md:3874:              searchParams.get("redirectTo") ?? undefined
./docs/tutorials/jokes.md:5254:              searchParams.get("redirectTo") ?? undefined
./docs/tutorials/jokes.md:5579:    request.headers.get("X-Forwarded-Host") ??
./scripts/playground/template/app/routes/join.tsx:80:  let redirectTo = searchParams.get("redirectTo") ?? undefined;
./scripts/release/github.ts:334:  let newNodes = res?.resource?.closingIssuesReferences?.nodes ?? [];
./packages/remix-react/components.tsx:1016:  let resolvedPath = useResolvedPath(action ?? ".");
./packages/remix-dev/cli/getPreferredPackageManager.ts:11:  ((process.env.npm_config_user_agent ?? "").split("/")[0] ||
./packages/remix-dev/cli/run.ts:220:  flags.interactive = flags.interactive ?? require.main === module;
./packages/remix-dev/cli/run.ts:418:      let useTypeScript = typescript ?? answers.useTypeScript;
./packages/remix-dev/__tests__/fixtures/replace-remix-imports/app/routes/join.tsx:83:  const redirectTo = searchParams.get("redirectTo") ?? undefined;
./packages/remix-dev/__tests__/fixtures/replace-remix-imports/app/routes/healthcheck.tsx:7:    request.headers.get("X-Forwarded-Host") ?? request.headers.get("host");
./packages/remix-dev/compiler.ts:413:    config.serverBuildTarget ?? ""
```

**Reply by sergiodxa** [reply]
That should be fixed IMO, so Remix can keep supporting any ESM compatible browser, or at least document supported browsers which may be a good idea anyway with a table like in MDN

**Reply by machour** [reply]
Opened https://github.com/remix-run/remix/pull/4561 that gets rid of the `./packages/remix-react/components.tsx` occurence, as it broke my website on an iPhone 6

**Reply by machour** [reply]
@qianjiachun if you're still having an issue, you can apply [this patch](https://gist.github.com/machour/870705998b8e14a96171ee2fb85d5921) using [patch-package](https://www.npmjs.com/package/patch-package)

**Reply by qianjiachun** [reply]
> @qianjiachun if you're still having an issue, you can apply [this patch](https://gist.github.com/machour/870705998b8e14a96171ee2fb85d5921) using [patch-package](https://www.npmjs.com/package/patch-package)

ok! I will try it.
thanks

**Reply by qianjiachun** [reply]
> @qianjiachun if you're still having an issue, you can apply [this patch](https://gist.github.com/machour/870705998b8e14a96171ee2fb85d5921) using [patch-package](https://www.npmjs.com/package/patch-package)

@machour 
I use `npm run dev` with your patch, and it was really worked! 👍
But when i use `npm run build` & `npm run start`, the `??` still exists.😂

I did the following:
1. install patch-package
2. copy your patch in `patches`
3. add `"postinstall": "patch-package"` into `scripts`
4. remove `node_modules` and npm install

I'm not sure if my steps are wrong, please give me some advice🙏

**Reply by machour** [reply]
Weird. Could you try removing the `.cache` folder and building again?
Also, can you search and see where's the `??` in the files after building? 

I only found and patched `node_modules/@remix-run/react/dist/components.js` and `node_modules/@remix-run/react/dist/esm/components.js` files, but maybe I missed one.

Also, I think 1.7.6 will be released today with the fix, so maybe you could wait for that instead

**Reply by qianjiachun** [reply]
Okay, i tried to remove `.cache` and rebuild, it was still failed😥.
Nevermind, i'm waiting for 1.7.6. Thanks again.

**Reply by machour** [reply]
@qianjiachun 1.7.6 is out, enjoy 🎉

**Reply by qianjiachun** [reply]
> @qianjiachun 1.7.6 is out, enjoy 🎉

Thank you for paying so much attention to this problem !

Although the problem still exists after `npm run build` (but it works in the `dev` environment), I checked `public/build/_ shared`, some codes include `??`. I think this is my own problem, perhaps caused by some modules I imported.

This problem is not very serious. I will do it myself, It doesn't matter. Thanks again.❤️❤️

**Reply by machour** [reply]
@qianjiachun You can ask for help on our [Discord](https://rmx.as/discord) !

**Comment by brophdawg11** [comment]
Hey folks - if you're looking to support older browsers, go give this proposal an upvote!  https://github.com/remix-run/remix/discussions/4769
